porthole/components.py

```diff
@@ -170,6 +172,63 @@ def create_worksheet_from_query(self, cm, sheet_name, query=None, sql=None, quer
         self.make_worksheet(sheet_name=sheet_name, query_results=results, **worksheet_kwargs)
 
 
+class ReportReadWriter(Loggable):
+
+    def __init__(self, report_filename: str
+                    , read_file_path: Optional[str]=None
+                    , write_file_path: Optional[str]=None
+                    , log_to=None):
+        self.report_filename = report_filename
+        if not read_file_path:
+            self.read_file_path = config['Default'].get('base_file_path')
+        if not write_file_path:
+            self.write_file_path = self.read_file_path
+        self.read_location = os.path.join(self.read_file_path, self.report_filename)
+        self.write_location = os.path.join(self.write_file_path, self.report_filename)
+        self.record_count = 0
+        super().__init__(log_to=log_to if log_to is not None else [])
+
+    def update_sheet_on_report_workbook(self, cm: ConnectionManager
+                                        , sheet_name: str
+                                        , query: Optional[dict]=None
+                                        , sql: Optional[str]=None
+                                        , query_kwargs: Optional[dict]=None
+                                        , save_as: Optional[str]=None) -> None:
+        if query is None:
+            query = {}
+        if query_kwargs is None:
+            query_kwargs = {}
+        if save_as:
+            self.write_location = os.path.join(self.write_file_path, save_as)
+        
+        results = self.execute_query(cm=cm, query=query, sql=sql, **query_kwargs)
+        formatted_results = results.as_list_of_tuples()
+        wb = WorkbookEditor(workbook_filename=self.read_location)
```

Comment:
This looks good to me. However, one thing I am wondering is whether it makes sense to open/close the same file potentially several times if updating multiple sheets. Would it be too much trouble to move the open and close of the file to separate methods? I might be over-optimizing, just wondering your thoughts.

Reply:
We would also need an `open_workbook_from_file` method.  Then we would modify this method to take a workbook object as a parameter.  Then we could break out the `close_workbook` method.  It would probably be more performant, at the cost of a couple extra lines in the implementation and remembering to manually open and close the workbook each time.  I don't have strong opinions on this, but am personally fine leaving it as-is for now.

Reply:
Being able to pass the workbook, rather than just a filename, makes testing easier.  So I will split apart this method.